### ISSUE TITLE
onready variable is not initialized inside _ready() of base class

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#33620]
**Godot version:**
3.2 beta1

**Issue description:**
This one is complicated a bit, but boils down to this:
```
extends Node2D
class_name Base

func _ready() -> void:
	test_node()

func test_node():
	pass
```

```
extends Base

onready var control = $Control

func test_node():
	print(control.name)
```

The `control` will be null in this case, which is a bit unexpected. Looks like the base script is initialized and calls `_ready` before the inheriting script is properly initialized.

**Minimal reproduction project:**
[ReproductionProject.zip](https://github.com/godotengine/godot/files/3846779/ReproductionProject.zip)

**Comment by groud** [comment]
I guess this is expected. The `onready` keywork is equivalent to setting the variable at the top of the `_ready` function. So when your script runs it initialize things in this order:

- run the parent Base's `_ready()` function 
- initialize the child's `onready` variable
- run the child `_ready()` function 

The the variable cannot be initialized inside the parent's call to the `_ready` function

**Comment by FrischerHering** [comment]
I also would say that this is expected behaviour, similar to C++ constructor call ordering with inherited objects.

You can solve this by calling `test_node()` from the child's `_ready()` function. Another way I did not test might be using Object's `call_deferred()` [see here](https://docs.godotengine.org/en/3.1/classes/class_object.html#class-object-method-call-deferred).

**Comment by Xrayez** [comment]
@KoBeWi I've managed to workaround the issue like that:
```gdscript
extends Base

#onready var control = $Control
var control

func _notification(what):
	match what:
		NOTIFICATION_INSTANCED:
			control = $Control
#		NOTIFICATION_READY: # same as onready
#			control = $Control

func test_node():
	print(control.name)

```

So basically the whole initialization happens on early `NOTIFICATION_INSTANCED` instead of syntax sugar coated `NOTIFICATION_READY`.

I've been thinking on migrating to this kind of technique over using `onready var` in fact which seems to resolve `onready` dependent node referencing with `get_node()` later on (while the scene is not even added to the tree). Talking about dependency: #30460.

---
I had to modify the scene tree because of wrong hierarchy of nodes, updated repro project:
[ReproductionProject.zip](https://github.com/godotengine/godot/files/3907371/ReproductionProject.zip)

**Comment by KoBeWi** [comment]
Still valid in 3.2.3
There's a chance it's fixed in master, but can't test due to #42440 and #42506

**Comment by KoBeWi** [comment]
Checked again in 4d08a73, seems like `onready` variable counts as if `_ready()` was defined, so the base method is not called. Not sure if this is intended.

**Comment by jakefriend-dev** [comment]
While I would understand why this was described as expected behaviour, I'd personally argue why this order would be preferable:

- Base class onready vars get set up
- Extending class onready vars get set up (and so on, for any subsequent extenders)
- Base class _ready() fires
- Extending class _ready() fires (and so on, for any subsequent extenders)

I imagine there's probably some kind of traditional best-practice reason why this isn't done presently, but in terms of use cases, I think it's in line with expected user behaviour to have a base class define signals that are emitted and for extending classes to be able to connect to those signals. So EG, if the base class fires a signal in its ready script, and the extending script's response involves an onready var Object reference doing something, you'll experience a crash because your references are still pointing to null.

Basically, I need to trust as an end user that "ready" means "ready" and not "somewhat ready". Since no one is going to vouch for adding an addition virtual function called `_ready_for_real()`, is there some solution that doesn't involve users creating awkward workarounds that can be actioned for this ticket, which would have the net effect of "when a base class ready signal fires, extending classes' onready vars are already set, so it is actually fully ready"?

**Comment by bsil78** [comment]
Just to add more to this, any _ready should just be called only when all is _ready ; thus according to @jakefriend-dev ; all onready that comes before should be called before, as awaited. However base class _ready should not be called if overriden by subclass, but is subclass call it itself by ._ready() as usual.

**Comment by YuriSizov** [comment]
I agree with the suggestion by Jake that `onready` setters should be its own pass before the `_ready` calls. Even in 4.0 where you have to be explicit, IIRC, with super class calls this would be a positive change for users without much downside.

For now, though, the simplest workaround is to `call_deferred()` in the super `_ready` method for any class method that can be overridden. It should be enough to give time for the extending scripts to run their ready logic.

**Comment by KoBeWi** [comment]
Fate of this issue depends on how #52223 is resolved.
Right now the issue is basically fixed, since you can do `super()` in your `_ready()`, which will initialize parent's `@onready` variables where you want it. But this is result of a bug.

#61550 initializes all `@onready` variables before any `_ready()` is called, but it will be superseded.
